File: packages/core/projects/collab-path.ts
// Client-side mirror of the server's collab_path validation
// (server/internal/handler/collab_path.go). The path is a project-level field
// — a module's deliverables live in a folder named after it inside the
// project's space, with no separate value to validate — so this lives beside
// the project contract rather than inside one UI surface; every project
// surface (web, desktop) imports it from here.
//
// The point of validating here is not to replace the server check — the server
// still answers 400 and is the authority — but to name the one mistake users
// actually make (typing a path relative to their own machine) in the field
// where they made it, instead of surfacing a raw API error in a toast.

/** Maximum stored length, in UTF-8 bytes. The server bounds `len(trimmed)`,
 *  which counts bytes, so a Chinese path hits the ceiling roughly three times
 *  sooner than its character count suggests. Counting characters here would
 *  let a value through that the server then rejects. */
export const COLLAB_PATH_MAX_BYTES = 1024;

export type CollabPathError =
  /** Longer than COLLAB_PATH_MAX_BYTES once trimmed. */
  | "too_long"
  /** Contains a control character other than TAB. */
  | "control_characters"
  /** Not a POSIX, UNC or Windows-drive absolute path. */
  | "not_absolute";

export type CollabPathResult =
  /** `value` is the trimmed path, or null when the input clears the field. */
  | { ok: true; value: string | null }
  | { ok: false; reason: CollabPathError };

/** UTF-8 byte length without allocating an encoder: keeps this module free of
 *  any platform global so it stays usable from a node test, a browser and a
 *  React Native bundle alike. */
function utf8ByteLength(value: string): number {
  let bytes = 0;
  for (const char of value) {
    const code = char.codePointAt(0) ?? 0;
    if (code <= 0x7f) bytes += 1;
    else if (code <= 0x7ff) bytes += 2;
    else if (code <= 0xffff) bytes += 3;
    else bytes += 4;
  }
  return bytes;
}

/** TAB is the one control character the server tolerates (a path may legally
 *  contain one); everything else in Unicode category Cc is a paste accident. */
const FORBIDDEN_CONTROL = /[\p{Cc}]/u;

/**
 * Absolute forms a daemon host can mount, matching isAbsoluteCollabPath on the
 * server: POSIX (`/Volumes/...`), UNC (`\\nas\share\...`) and a Windows drive
 * (`Z:/...` or `Z:\...`). Daemons run on macOS, Linux and Windows, so pinning
 * the check to one separator would reject a value that is correct on another.
 */
export function isAbsoluteCollabPath(path: string): boolean {
  if (path.startsWith("/") || path.startsWith("\\\\")) return true;
  if (path.length >= 3 && path[1] === ":" && (path[2] === "/" || path[2] === "\\")) {
    return /^[A-Za-z]$/.test(path[0] ?? "");
  }
  return false;
}

/**
 * Trims and validates a human-agent collaboration space path
 * ("人机协作空间路径").
 *
 * A blank value means "no path" and resolves to null, which callers send as an
 * explicit null to clear the column — the update contract reads an absent key
 * as "keep the current value", so a cleared field must still be sent.
 */
export function normalizeCollabPath(raw: string): CollabPathResult {
  const trimmed = raw.trim();
  if (trimmed === "") return { ok: true, value: null };
  if (utf8ByteLength(trimmed) > COLLAB_PATH_MAX_BYTES) {
    return { ok: false, reason: "too_long" };
  }
  if (FORBIDDEN_CONTROL.test(trimmed.replace(/\t/g, ""))) {
    return { ok: false, reason: "control_characters" };
  }
  if (!isAbsoluteCollabPath(trimmed)) {
    return { ok: false, reason: "not_absolute" };
  }
  return { ok: true, value: trimmed };
}

// ---------------------------------------------------------------------------
// Addressing the same directory from a reader's machine
// ---------------------------------------------------------------------------

/**
 * A collaboration space is stored as the path the AGENT's machine mounts, and
 * that string is useless to a reader's browser: an http(s) page may not
 * navigate to `file://`, in any browser, with no permission a user can grant.
 *
 * The directory itself is reachable, though — it is on a file server both
 * machines can see. Given that server's host (the deployment's
 * `collab_space_host`), the same location has two addresses that do work:
 *
 *   smb://host/share/…   macOS hands this to Finder, so a click opens it
 *   \\host\share\…       what Windows Explorer accepts in its address bar
 *
 * Deriving them requires knowing which part of the path is the share name.
 * Only two forms answer that. A UNC path carries the host and share itself, so
 * it needs no configuration at all. And macOS mounts a share at
 * `/Volumes/<share name>`, so the first segment under `/Volumes` IS the share.
 *
 * Everything else returns nothing, deliberately. A Linux `/mnt/<anything>` is
 * named by whoever wrote the fstab entry and a `Z:\…` drive letter hides the
 * share behind a per-machine mapping; guessing either would produce an address
 * that fails in a way the reader cannot diagnose, which is worse than the
 * clipboard.
 */
export interface CollabPathAddresses {
  /** Opens in Finder on macOS. `null` when the share cannot be identified. */
  smbUrl: string | null;
  /** Paste-able in the Windows Explorer address bar. */
  uncPath: string | null;
}

const NONE: CollabPathAddresses = { smbUrl: null, uncPath: null };

const MACOS_MOUNT_ROOT = "/Volumes/";

/** Percent-encodes each segment while leaving the separators alone. macOS
 *  writes its own mount URLs this way, so Finder decodes them back. */
function encodeSmbPath(path: string): string {
  return path
    .split("/")
    .map((segment) => encodeURIComponent(segment))
    .join("/");
}

export function collabPathAddresses(
  path: string,
  host: string,
): CollabPathAddresses {
  const value = path.trim();
  if (!value) return NONE;

  // A UNC path already names its own server. Nothing to configure, and the
  // deployment's host is irrelevant — this path may well name a different one.
  if (value.startsWith("\\\\")) {
    const rest = value.slice(2).replace(/\\/g, "/");
    const [uncHost, ...segments] = rest.split("/").filter(Boolean);
    // `\\host` alone is a server, not a directory.
    if (!uncHost || segments.length === 0) return NONE;
    return {
      smbUrl: `smb://${uncHost}/${encodeSmbPath(segments.join("/"))}`,
      uncPath: value.replace(/\//g, "\\"),
    };
  }

  const server = host.trim();
  if (!server) return NONE;
  if (!value.startsWith(MACOS_MOUNT_ROOT)) return NONE;
  const rest = value.slice(MACOS_MOUNT_ROOT.length);
  // `/Volumes/` alone names the mount root, not a share.
  if (!rest || rest.startsWith("/")) return NONE;

  return {
    smbUrl: `smb://${server}/${encodeSmbPath(rest)}`,
    uncPath: `\\\\${server}\\${rest.replace(/\//g, "\\")}`,
  };
}

File: server/internal/handler/collab_path.go
package handler

import (
	"fmt"
	"net/http"
	"strings"
	"unicode"

	"github.com/jackc/pgx/v5/pgtype"
)

// collabPathMaxLen bounds the stored path. Deep NAS trees are legitimate, but a
// value this long is a paste accident, not a directory.
const collabPathMaxLen = 1024

// normalizeCollabPath validates and trims a human-agent collaboration space
// path ("人机协作空间路径"): the directory on shared storage where a project
// exchanges deliverables between people and agents. Only a project stores one —
// a module's folder sits inside it under the module's own name.
//
// The server never stats the path. It is resolved on whichever daemon host
// runs the task, and a host that has not mounted the share must fail loudly at
// the agent rather than have the platform reject a value that is correct
// everywhere else. So validation only rejects values that cannot be a path on
// any host: relative paths (which would silently resolve inside a task's
// private workdir — the one place a deliverable must not land), control
// characters, and absurd lengths.
//
// A blank value after trimming means "no path", which callers store as NULL.
func normalizeCollabPath(raw string) (pgtype.Text, error) {
	trimmed := strings.TrimSpace(raw)
	if trimmed == "" {
		return pgtype.Text{Valid: false}, nil
	}
	if len(trimmed) > collabPathMaxLen {
		return pgtype.Text{}, fmt.Errorf("collab_path must be at most %d characters", collabPathMaxLen)
	}
	for _, r := range trimmed {
		if r == '\x00' || (unicode.IsControl(r) && r != '\t') {
			return pgtype.Text{}, fmt.Errorf("collab_path must not contain control characters")
		}
	}
	if !isAbsoluteCollabPath(trimmed) {
		return pgtype.Text{}, fmt.Errorf("collab_path must be an absolute path (for example /Volumes/人机协作空间/项目/模块)")
	}
	return pgtype.Text{String: trimmed, Valid: true}, nil
}

// isAbsoluteCollabPath accepts the absolute forms a daemon host can mount:
// POSIX ("/Volumes/..."), Windows drive ("Z:\..." or "Z:/...") and UNC
// ("\\nas\share\..."). Daemons run on macOS, Linux and Windows, so pinning the
// check to one separator would reject a correct value on the other platforms.
func isAbsoluteCollabPath(p string) bool {
	if strings.HasPrefix(p, "/") || strings.HasPrefix(p, `\\`) {
		return true
	}
	if len(p) >= 3 && p[1] == ':' && (p[2] == '/' || p[2] == '\\') {
		c := p[0]
		return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
	}
	return false
}

// collabPathFromRequest resolves an optional request field into column state,
// writing a 400 and returning ok=false when the value is unusable. A nil
// pointer clears the column, matching the "present key, null value" contract
// the project and module update handlers use for every other nullable field.
func collabPathFromRequest(w http.ResponseWriter, value *string) (pgtype.Text, bool) {
	if value == nil {
		return pgtype.Text{Valid: false}, true
	}
	normalized, err := normalizeCollabPath(*value)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return pgtype.Text{}, false
	}
	return normalized, true
}
